packages/muelu/src/Transfers/GeneralGeometric/MueLu_GeneralGeometricPFactory_decl.hpp

```diff
@@ -0,0 +1,150 @@
+  | LineDetection_Layers | Factory | null | | * | * | Generating factory for LineDetection information. Usually provided by the LineDetectionFactory. Array with layer id for all nodes on current processor. |
+  | CoarseNumZLayers | Factory | null | | * | * | Generating factory for LineDetection information. Usually provided by the LineDetectionFactory. Number of remaining z-layers after semi-coarsening. |
+  | semicoarsen: coarsen rate | int | * | * |   | Coarsening rate along vertical lines (2 corresponds to classical semicoarsening. Values > 2 for more aggressive coarsening). |
+
+
+
+  The * in the @c master.xml column denotes that the parameter is defined in the @c master.xml file.<br>
+  The * in the @c validated column means that the parameter is declared in the list of valid input parameters (see SemiCoarsenPFactory::GetValidParameters).<br>
+  The * in the @c requested column states that the data is requested as input with all dependencies (see SemiCoarsenPFactory::DeclareInput).
+
+  ### Variables provided by SemiCoarsenPFactory ###
+
+  After SemiCoarsenPFactory::Build the following data is available (if requested)
+
+  Parameter | generated by | description
+  ----------|--------------|------------
+  | P       | SemiCoarsenPFactory   | Prolongator
+  | Nullspace | SemiCoarsenPFactory | Coarse nullspace (the fine level nullspace information is coarsened using P to generate a coarse version of the nullspace. No scaling is applied.
+  | NumZLayers | NoFactory | Number of z layers after coarsening. Necessary input for LineDetectionFactory. Useful input for TogglePFactory.
+
+*/
+  template <class Scalar = double, class LocalOrdinal = int, class GlobalOrdinal = LocalOrdinal, class Node = KokkosClassic::DefaultNode::DefaultNodeType>
+  class GeneralGeometricPFactory : public PFactory {
+#undef MUELU_GENERALGEOMETRICPFACTORY_SHORT
+#include "MueLu_UseShortNames.hpp"
+
+  public:
+    //! @name Constructors/Destructors.
+    //@{
+
+    //! Constructor
+    GeneralGeometricPFactory() : bTransferCoordinates_(false) { }
+
+    //! Destructor.
+    virtual ~GeneralGeometricPFactory() { }
+    //@}
+
+    RCP<const ParameterList> GetValidParameterList() const;
+
+    //! Input
+    //@{
+
+    void DeclareInput(Level& fineLevel, Level& coarseLevel) const;
+
+    //@}
+
+    //! @name Build methods.
+    //@{
+
+    void Build (Level& fineLevel, Level& coarseLevel) const;
+    void BuildP(Level& fineLevel, Level& coarseLevel) const;
+
+    //@}
+
+  private:
+    // LO FindCpts(LO const PtsPerLine, LO const CoarsenRate, LO const Thin, LO **LayerCpts) const;
+    LO MakeGeneralGeometricP(LO const ndm, LO const fNodePerDir[], LO const cNodePerDir[], LO const coarse_rate[], LO const end_rate[],
+			     const RCP<Xpetra::MultiVector<double,LO,GO,NO> > fCoords, LO const nnzP, LO const dofsPerNode, RCP<const Map>& coarseMap,
```

Comment:
Yes, fCoords should passed to MakeGeneralGeometricP as RCP<>&, this is fixed now